staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -476,37 +478,35 @@ func (f *DeltaFIFO) IsClosed() bool {
 func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 	f.lock.Lock()
 	defer f.lock.Unlock()
-	for {
-		for len(f.queue) == 0 {
-			// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
-			// When Close() is called, the f.closed is set and the condition is broadcasted.
-			// Which causes this loop to continue and return from the Pop().
-			if f.closed {
-				return nil, ErrFIFOClosed
-			}
-
-			f.cond.Wait()
-		}
-		id := f.queue[0]
-		f.queue = f.queue[1:]
-		if f.initialPopulationCount > 0 {
-			f.initialPopulationCount--
+	for len(f.queue) == 0 {
+		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
+		// When Close() is called, the f.closed is set and the condition is broadcasted.
+		// Which causes this loop to continue and return from the Pop().
+		if f.closed {
+			return nil, ErrFIFOClosed
 		}
-		item, ok := f.items[id]
-		if !ok {
-			// Item may have been deleted subsequently.
-			continue
-		}
-		delete(f.items, id)
-		err := process(item)
-		if e, ok := err.(ErrRequeue); ok {
-			f.addIfNotPresent(id, item)
-			err = e.Err
-		}
-		// Don't need to copyDeltas here, because we're transferring
-		// ownership to the caller.
-		return item, err
+
+		f.cond.Wait()
+	}
+	id := f.queue[0]
+	f.queue = f.queue[1:]
+	if f.initialPopulationCount > 0 {
+		f.initialPopulationCount--
 	}
+	item, ok := f.items[id]
+	if !ok {
+		// This should never happen
+		panic(id)
+	}
+	delete(f.items, id)
+	err := process(item)
+	if e, ok := err.(ErrRequeue); ok {
+		f.addIfNotPresent(id, item)
```

Comment:
Out of scope of this PR, but I've always been wondering, what is the point of calling `addIfNotPresent` in deltaFIFO. The entire `pop()` operation was invoked under lock (acquired after `cond.wait()`), and we deleted the key from both `queue` and `items`, and there's no way the key would still exist after `process()`, right?